File: goofy-protocol-fis-be/src/main/java/com/masl/goofy_protocol_core/crypto/connected/HandleCrypto.java
package com.masl.goofy_protocol_core.crypto.connected;

import com.masl.goofy_protocol_core.crypto.isolated.SecretUtils;
import com.masl.goofy_protocol_core.crypto.isolated.asymm.AsymmCrypto;
import com.masl.goofy_protocol_core.crypto.isolated.asymm.GlobAsymmCrypto;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Pattern;

public class HandleCrypto implements GenericHandleCrypto {
    private final GlobAsymmCrypto asymmCrypto = new GlobAsymmCrypto();
    private final HandleCryptoHelper handleCryptoHelper;
    private final List<String> wordList;

    private Map<String, String> userKeyToHandleCache;
    private Map<String, String> generalKeyToHandleCache;
    private Map<String, String> sharedHandleToKeyCache;

    public record HandleCryptoConfig() {}

    public HandleCrypto(HandleCryptoHelper handleCryptoHelper) {
        this(handleCryptoHelper, new HandleCryptoConfig());
    }

    public HandleCrypto(HandleCryptoHelper handleCryptoHelper, HandleCryptoConfig config) {
        this.handleCryptoHelper = handleCryptoHelper;

        wordList = handleCryptoHelper.loadWordList();
        generalKeyToHandleCache = new ConcurrentHashMap<>(handleCryptoHelper.loadPersistedKeyToHandleMapCache());
        userKeyToHandleCache = new ConcurrentHashMap<>(handleCryptoHelper.loadUserKeyToHandleMap());
        generateHandleToKeyMapping();
    }

    // Persistence

    synchronized public void reloadUserKeyToHandle() {
        userKeyToHandleCache = new ConcurrentHashMap<>(handleCryptoHelper.loadUserKeyToHandleMap());
        generateHandleToKeyMapping();
    }

    synchronized public void reloadGeneralKeyToHandle() {
        generalKeyToHandleCache = new ConcurrentHashMap<>(handleCryptoHelper.loadPersistedKeyToHandleMapCache());
        generateHandleToKeyMapping();
    }

    synchronized public void saveGeneralKeyToHandle() {
        handleCryptoHelper.storePersistedKeyToHandleMapCache(generalKeyToHandleCache);
    }

    // Generate Reverse Mappings

    synchronized public void generateHandleToKeyMapping() {
        sharedHandleToKeyCache = new ConcurrentHashMap<>();
        userKeyToHandleCache.forEach((pubSplitKey, handle) ->
                sharedHandleToKeyCache.putIfAbsent(handle, pubSplitKey));
        generalKeyToHandleCache.forEach((pubSplitKey, handle) ->
                sharedHandleToKeyCache.putIfAbsent(handle, pubSplitKey));
    }

    // Adding Entries

    public void addUserKeyToHandle(String pubSplitKey, String handle) {
        if (!verifyKeyAndHandle(pubSplitKey, handle, false))
            throw new IllegalArgumentException("Invalid key-handle pair");

        userKeyToHandleCache.putIfAbsent(pubSplitKey, handle);
        sharedHandleToKeyCache.putIfAbsent(handle, pubSplitKey);
    }

    public void addGeneralKeyToHandle(String pubSplitKey, String handle) {
        if (!verifyKeyAndHandle(pubSplitKey, handle, false))
            throw new IllegalArgumentException("Invalid key-handle pair");

        generalKeyToHandleCache.putIfAbsent(pubSplitKey, handle);
        sharedHandleToKeyCache.putIfAbsent(handle, pubSplitKey);
    }

    // Verification

    public boolean verifyKeyAndHandle(String pubSplitKey, String handle) {
        return verifyKeyAndHandle(pubSplitKey, handle, true);
    }

    public boolean verifyKeyAndHandle(AsymmCrypto.AsymmPubKeyPair pubSplitKey, String handle) {
        return verifyKeyAndHandle(pubSplitKey.serialize(), handle, true);
    }


    public boolean verifyKeyAndHandle(String pubSplitKey, String handle, boolean useCache) {
        if (useCache)
            return deriveHandle(pubSplitKey).equals(handle);
        return _internalDeriveHandle(pubSplitKey).equals(handle);
    }

    // Keypair to Handle Derivation

    public String deriveHandle(AsymmCrypto.AsymmPubKeyPair pubSplitKey) {
        return deriveHandle(pubSplitKey.serialize());
    }

    public String deriveHandle(String pubSplitKey) {
        if (userKeyToHandleCache.containsKey(pubSplitKey))
            return userKeyToHandleCache.get(pubSplitKey);

        return generalKeyToHandleCache.computeIfAbsent(pubSplitKey, _pubSplitKey -> {
            String handle = _internalDeriveHandle(_pubSplitKey);
            if (handle != null)
                sharedHandleToKeyCache.putIfAbsent(handle, _pubSplitKey);
            return handle;
        });
    }

    synchronized public String _internalDeriveHandle(String pubSplitKey) {
        // Check if it can be parsed / is valid
        AsymmCrypto.AsymmPubKeyPair pair = AsymmCrypto.AsymmPubKeyPair.parse(pubSplitKey);
        if (!pair.isSigValid(asymmCrypto.forType(pair.type())))
            throw new IllegalArgumentException("Invalid pubSplitKey");

        // Strength of handles
        // c = 2 -> ~44 bit (15000^2 * 10^5 = 2.3e13)
        // c = 3 -> ~58 bit (15000^3 * 10^5 = 3.4e17)
        // c = 4 -> ~72 bit (15000^4 * 10^5 = 5.1e21)

        byte[] rootRnd = SecretUtils.symmSecretFromSecret(pubSplitKey, SecretUtils.DEFAULT_HANDLE_ROOT_SALT, 128, SecretUtils.DEFAULT_HANDLE_ROOT_ITERATIONS);
        if (rootRnd == null)
            throw new IllegalArgumentException("Invalid pubSplitKey (unable to derive rootRnd)");
        int _c = rootRnd[0] & 0xFF;
        int _n = ((rootRnd[1] & 0xFF) << 8) | (rootRnd[2] & 0xFF);

        // Sanity Check
        if (_c < 0 || _n < 0)
            throw new IllegalArgumentException("Invalid pubSplitKey (negative c or n)");

        // Put into Bounds
        int c = 2 + _c % 3;
        int n = _n % 100000;

        // Select and write the words
        byte[] lastBytes = rootRnd;
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < c; i++) {
            String newSeed = Base64.getEncoder().encodeToString(lastBytes);
            byte[] wordRnd = SecretUtils.symmSecretFromSecret(newSeed, SecretUtils.DEFAULT_HANDLE_WORD_SALT, 128, SecretUtils.DEFAULT_HANDLE_WORD_ITERATIONS);
            if (wordRnd == null)
                throw new IllegalArgumentException("Invalid pubSplitKey (unable to derive wordRnd)");
            lastBytes = wordRnd; // use as base for next word

            int _wordIdx = ((wordRnd[0] & 0xFF) << 8) | (wordRnd[1] & 0xFF);
            int wordIndex = _wordIdx % wordList.size();

            String word = wordList.get(wordIndex);
            builder.append(word);
            if (i < c - 1)
                builder.append("_");
        }

        // Append number
        builder.append(n);
        return builder.toString();
    }

    // Handle & Domain Separation

    public String stripPotentialDomainFromHandle(String handle) {
        return handle.split(Pattern.quote(DEF_DOMAIN_SEPARATOR))[0];
    }

    public String getPotentialDomainFromHandle(String handle) {
        String[] split = handle.split(Pattern.quote(DEF_DOMAIN_SEPARATOR));
        if (split.length > 1)
            return split[1];
        return null;
    }

    // Handle to Keypair Lookup

    public String getPublicSplitKeyFromHandle(String handle) {
        String strippedHandle = stripPotentialDomainFromHandle(handle);
        return sharedHandleToKeyCache.computeIfAbsent(strippedHandle, _ -> {
            String pubSplitKey = handleCryptoHelper.lookupPubSplitKeyForHandleExternally(handle);
            if (pubSplitKey != null)
                generalKeyToHandleCache.putIfAbsent(pubSplitKey, strippedHandle);
            return pubSplitKey;
        });
    }
}

File: goofy-protocol-fis-fe/src/libs/goofy-protocol-core/crypto/connected/handle-crypto.mjs
import { AsymmPubKeyPair } from "../isolated/asymm/asymm-crypto-interface.mjs";
import { GlobAsymmCrypto } from "../isolated/asymm/glob-asymm-crypto.mjs";
import {
    symmSecretFromSecret,
    DEFAULT_HANDLE_ROOT_ITERATIONS,
    DEFAULT_HANDLE_WORD_ITERATIONS,
    DEFAULT_HANDLE_ROOT_SALT,
    DEFAULT_HANDLE_WORD_SALT,
} from "../isolated/secret-utils.mjs";


// ---- HandleCrypto ----

export class HandleCrypto {
    /**
     * @param {object} handleCryptoHelper
     * @param {object} [config]
     */
    constructor(handleCryptoHelper, config = {}) {
        this.asymmCrypto = new GlobAsymmCrypto();
        this.handleCryptoHelper = handleCryptoHelper;

        this.wordList = handleCryptoHelper.loadWordList();

        // Maps: key->handle and handle->key
        this.generalKeyToHandleCache = new Map(handleCryptoHelper.loadPersistedKeyToHandleMapCache());
        this.userKeyToHandleCache = new Map(handleCryptoHelper.loadUserKeyToHandleMap());

        this.sharedHandleToKeyCache = new Map();
        this.generateHandleToKeyMapping();
    }

    // Persistence
    reloadUserKeyToHandle() {
        this.userKeyToHandleCache = new Map(this.handleCryptoHelper.loadUserKeyToHandleMap());
        this.generateHandleToKeyMapping();
    }

    reloadGeneralKeyToHandle() {
        this.generalKeyToHandleCache = new Map(this.handleCryptoHelper.loadPersistedKeyToHandleMapCache());
        this.generateHandleToKeyMapping();
    }

    // Generate reverse mappings
    generateHandleToKeyMapping() {
        this.sharedHandleToKeyCache = new Map();

        for (const [pubSplitKey, handle] of this.userKeyToHandleCache.entries()) {
            if (!this.sharedHandleToKeyCache.has(handle)) this.sharedHandleToKeyCache.set(handle, pubSplitKey);
        }
        for (const [pubSplitKey, handle] of this.generalKeyToHandleCache.entries()) {
            if (!this.sharedHandleToKeyCache.has(handle)) this.sharedHandleToKeyCache.set(handle, pubSplitKey);
        }
    }

    // Adding Entries
    async addUserKeyToHandle(pubSplitKey, handle) {
        const ok = await this.verifyKeyAndHandle(pubSplitKey, handle, false);
        if (!ok) throw new Error("Invalid key-handle pair");

        this.userKeyToHandleCache.set(pubSplitKey, this.userKeyToHandleCache.has(pubSplitKey) ? this.userKeyToHandleCache.get(pubSplitKey) : handle);
        if (!this.sharedHandleToKeyCache.has(handle)) this.sharedHandleToKeyCache.set(handle, pubSplitKey);
    }

    async addGeneralKeyToHandle(pubSplitKey, handle) {
        const ok = await this.verifyKeyAndHandle(pubSplitKey, handle, false);
        if (!ok) throw new Error("Invalid key-handle pair");

        this.generalKeyToHandleCache.set(pubSplitKey, this.generalKeyToHandleCache.has(pubSplitKey) ? this.generalKeyToHandleCache.get(pubSplitKey) : handle);
        if (!this.sharedHandleToKeyCache.has(handle)) this.sharedHandleToKeyCache.set(handle, pubSplitKey);
    }

    // Verification

    async verifyKeyAndHandleFromPair(pubSplitKeyPair, handle) {
        return this.verifyKeyAndHandle(pubSplitKeyPair.serialize(), handle, true);
    }

    async verifyKeyAndHandle(pubSplitKey, handle, useCache = true) {
        if (useCache) {
            const derived = await this.deriveHandle(pubSplitKey);
            return derived === handle;
        }
        const derived = await this._internalDeriveHandle(pubSplitKey);
        return derived === handle;
    }

    // Keypair to Handle Derivation
    async deriveHandle(pubSplitKeyOrPair) {
        const pubSplitKey =
            pubSplitKeyOrPair instanceof AsymmPubKeyPair ? pubSplitKeyOrPair.serialize() : pubSplitKeyOrPair;

        if (this.userKeyToHandleCache.has(pubSplitKey)) {
            return this.userKeyToHandleCache.get(pubSplitKey);
        }

        // JS equivalent of computeIfAbsent (async) with simple memoization
        if (this.generalKeyToHandleCache.has(pubSplitKey)) {
            return this.generalKeyToHandleCache.get(pubSplitKey);
        }

        const handle = await this._internalDeriveHandle(pubSplitKey);
        if (handle != null) {
            this.handleCryptoHelper.addPersistedKeyToHandleMapping(pubSplitKey, handle);
            if (!this.sharedHandleToKeyCache.has(handle)) this.sharedHandleToKeyCache.set(handle, pubSplitKey);
            this.generalKeyToHandleCache.set(pubSplitKey, handle);
        }
        return handle;
    }

    async _internalDeriveHandle(pubSplitKey) {
        const pair = AsymmPubKeyPair.parse(pubSplitKey);

        const cryptoImpl = this.asymmCrypto.forType(pair.type);
        if (!cryptoImpl) throw new Error("Invalid pubSplitKey");
        const sigOk = await pair.isSigValid(cryptoImpl);
        if (!sigOk) throw new Error("Invalid pubSplitKey");

        // Strength of handles
        const rootRnd = await symmSecretFromSecret(
            pubSplitKey,
            DEFAULT_HANDLE_ROOT_SALT,
            128,
            DEFAULT_HANDLE_ROOT_ITERATIONS
        );
        if (rootRnd == null) throw new Error("Invalid pubSplitKey (unable to derive rootRnd)");

        const _c = rootRnd[0] & 0xff;
        const _n = ((rootRnd[1] & 0xff) << 8) | (rootRnd[2] & 0xff);

        if (_c < 0 || _n < 0) throw new Error("Invalid pubSplitKey (negative c or n)");

        const c = 2 + (_c % 3);
        const n = _n % 100000;

        // Append words derived from successive PBKDF2 steps
        let lastBytes = rootRnd;
        const parts = [];
        for (let i = 0; i < c; i++) {
            const newSeed = base64Encode(lastBytes);
            const wordRnd = await symmSecretFromSecret(
                newSeed,
                DEFAULT_HANDLE_WORD_SALT,
                128,
                DEFAULT_HANDLE_WORD_ITERATIONS
            );
            if (wordRnd == null) throw new Error("Invalid pubSplitKey (unable to derive wordRnd)");

            lastBytes = wordRnd;

            const _wordIdx = ((wordRnd[0] & 0xff) << 8) | (wordRnd[1] & 0xff);
            const wordIndex = _wordIdx % this.wordList.length;

            parts.push(this.wordList[wordIndex]);
        }

        return `${parts.join("_")}${n}`;
    }

    // Handle to Keypair Lookup
    async getPublicSplitKeyFromHandle(handle) {
        const strippedHandle = this.constructor.stripPotentialDomainFromHandle(handle);

        if (this.sharedHandleToKeyCache.has(strippedHandle)) {
            return this.sharedHandleToKeyCache.get(strippedHandle);
        }

        const pubSplitKey = await this.handleCryptoHelper.lookupPubSplitKeyForHandleExternally(handle);

        if (pubSplitKey != null) {
            this.handleCryptoHelper.addPersistedKeyToHandleMapping(pubSplitKey, strippedHandle);
            this.generalKeyToHandleCache.set(pubSplitKey, strippedHandle);
            if (!this.sharedHandleToKeyCache.has(strippedHandle)) this.sharedHandleToKeyCache.set(strippedHandle, pubSplitKey);
        }

        return pubSplitKey;
    }

    // ---- Domain stripping (you had GenericHandleCrypto.stripPotentialDomainFromHandle) ----
    static stripPotentialDomainFromHandle(handle) {
        // If your JS environment has GenericHandleCrypto, you can replace this with the real call.
        // Common patterns: "handle@domain", "domain:handle", etc.
        // This is a safe default: return the substring before '@' if present.
        const at = handle.indexOf("@");
        return at >= 0 ? handle.slice(0, at) : handle;
    }
}

// ---- Base64 for seeds (Java used Base64.getEncoder().encodeToString(byte[])) ----
function base64Encode(bytes) {
    let binary = "";
    const chunkSize = 0x8000;
    for (let i = 0; i < bytes.length; i += chunkSize) {
        binary += String.fromCharCode(...bytes.subarray(i, i + chunkSize));
    }
    return btoa(binary);
}